src/hotstuff/protocol.rs

```diff
@@ -3,154 +3,910 @@
+            // views of voting for a validator-set-updating block proposed earlier has been interrupted.
+            if let Some(block_hash) = repropose_block(self.view_info.view, block_tree)? {
+                let block = block_tree
+                    .block(&block_hash)?
+                    .ok_or(BlockTreeError::BlockExpectedButNotFound { block: block_hash })?;
+
+                let proposal_msg =
+                    HotStuffMessage::proposal(self.config.chain_id, self.view_info.view, block);
+                self.sender_handle.broadcast(proposal_msg.clone());
+
+                if let HotStuffMessage::Proposal(proposal) = proposal_msg {
+                    Event::Propose(ProposeEvent {
+                        timestamp: SystemTime::now(),
+                        proposal,
+                    })
+                    .publish(&self.event_publisher)
+                }
+
+                return Ok(());
+            }
+
+            // Otherwise, propose a new block or nudge based on highest_qc.
+            let highest_qc = block_tree.highest_qc()?;
+            match highest_qc.phase {
+                // Produce and broadcast a new Proposal.
+                Phase::Generic | Phase::Decide => {
+                    let (parent_block, child_height) = if highest_qc.is_genesis_qc() {
+                        (None, BlockHeight::new(0))
+                    } else {
+                        let parent_height = block_tree.block_height(&highest_qc.block)?.ok_or(
+                            BlockTreeError::BlockExpectedButNotFound {
+                                block: highest_qc.block,
+                            },
+                        )?;
+                        (Some(highest_qc.block), parent_height + 1)
+                    };
 
-        // 1. Broadcast a NewView message for self.cur_view to leader.
+                    let produce_block_request = ProduceBlockRequest::new(
+                        self.view_info.view,
+                        parent_block,
+                        block_tree.app_view(parent_block.as_ref())?,
+                    );
 
-        // 2. Update cur_view and cur_leader to view and leader, next_leader tp next_leader.
+                    let ProduceBlockResponse {
+                        data,
+                        data_hash,
+                        app_state_updates: _,
+                        validator_set_updates: _,
+                    } = app.produce_block(produce_block_request);
 
-        // 3. If I am cur_leader then broadcast a nudge or a proposal.
+                    let block = Block::new(child_height, highest_qc, data_hash, data);
 
-        todo!()
+                    let proposal_msg =
+                        HotStuffMessage::proposal(self.config.chain_id, self.view_info.view, block);
+
+                    self.sender_handle.broadcast(proposal_msg.clone());
+
+                    if let HotStuffMessage::Proposal(proposal) = proposal_msg {
```

Comment:
This ugly pattern of using `if let` statements to decompose `HotStuffMessage`s to take out the inner, specific message type is used throughout this module. The `if` part here is clearly redundant since it is infallible: the code block will always be entered in every case.

I think we can consider removing this pattern by making the constructors of `Proposal`, `Vote`, and `NewView` inherent static methods of the specific message types, instead of inherent static methods of the `HotStuffMessage` type. 

This way, we can construct instances of the specific message types instead of having to construct instances of `HotStuffMessage` directly. Then, we can implement `Into<Message>` for each of the specific message types so that `SenderHandle::send` can accept